File: PoeLogsParser/PoeLogsParser/Services/LogReaderService.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PoeLogsParser.Exceptions;
using PoeLogsParser.Models;
using PoeLogsParser.Services.Abstractions;

namespace PoeLogsParser.Services
{
    public class LogReaderService : ILogReaderService
    {
        #region Events

        public delegate void NewLogEntryEvent(string line);

        public event NewLogEntryEvent OnNewLogEntry;

        #endregion

        #region Constants

        private readonly string[] _poeProcesses = new[]
        {
            "PathOfExile_x64",
            "PathOfExile_x64Steam"
        };

        #endregion

        #region Members

        private string _logFilePath;
        private bool _fileOpened = false;
        private FileStream _file;
        private long _endOfDile = 0;

        #endregion

        #region Constructors

        public LogReaderService(string logFilePath)
        {
            _logFilePath = logFilePath;
            Watch();
        }

        public LogReaderService()
        {
            FindLogFilePath();
            Watch();
        }

        #endregion

        private void FindLogFilePath()
        {
            var processes = Process.GetProcesses();

            foreach (var process in processes)
            {
                if (!_poeProcesses.Contains(process.ProcessName) || process.HasExited) continue;
                try
                {
                    if (process.MainModule?.FileName != null)
                        _logFilePath =
                            $"{(process.MainModule?.FileName)[..process.MainModule.FileName.LastIndexOf("\\", StringComparison.Ordinal)]}\\logs\\Client.txt";
                } catch (Exception e)
                {
                    throw new CannotFindLogFileException(process.ProcessName);
                }
            }
        }

        private async void Watch()
        {
            while (true)
            {
                var newLines = new List<string>();

                do
                {
                    try
                    {
                        await Task.Delay(500);

                        newLines = this.ReadNewLines();
                    } catch
                    {
                        // ignored
                    }
                } while (!newLines.Any());

                foreach (var line in newLines)
                {
                    OnOnNewLogEntry(line);
                }
            }
            // ReSharper disable once FunctionNeverReturns
        }

        private void SetEndOfFile()
        {
            var file = File.Open(_logFilePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            _endOfDile = file.Length - 1;
            file.Close();
        }

        private List<string> ReadNewLines()
        {
            var lines = new List<string>();

            var currentPosition = _endOfDile;

            SetEndOfFile();

            if (currentPosition >= _endOfDile)
            {
                return lines;
            }

            var file = File.Open(_logFilePath, FileMode.Open, FileAccess.Read,
                FileShare.ReadWrite);
            file.Position = currentPosition;
            var reader = new StreamReader(file);

            var line = reader.ReadLine();
            while (!reader.EndOfStream)
            {
                if (!string.IsNullOrEmpty(line))
                {
                    lines.Add(line);
                }
            }

            reader.Close();
            file.Close();

            return lines;
        }

        private bool OpenLogFile()
        {
            if (string.IsNullOrEmpty(_logFilePath))
            {
                return false;
            }

            FindLogFilePath();

           // new FileStream(@"c:\file.txt", FileMode.Open, FileAccess.Read)

            return true;
        }

        public string ReadLastLine()
        {
            throw new System.NotImplementedException();
        }

        public string ReadLine(int lineNo)
        {
            throw new System.NotImplementedException();
        }

        public IEnumerable<string> ReadLines(int[] linesNo)
        {
            throw new System.NotImplementedException();
        }


        protected virtual void OnOnNewLogEntry(string line)
        {
            OnNewLogEntry?.Invoke(line);
        }
    }
}

File: ts/src/services/LogReaderService.ts
const findProcess = require('find-process');
import { stat, open, read } from 'fs';
import { platform, EOL } from 'os'
import { Subject } from 'rxjs'

export class LogReaderService {
	private _poeProcesses: string[] = [
		"PathOfExile_x64",
		"PathOfExile_x64Steam"
	];
	private _logFilePath!: string;
	private _endOfFile: number = 0;

	public newLogEntry: Subject<string>;

	constructor(logFilePath: string | undefined = undefined) {
		this.newLogEntry = new Subject<string>();

		if (logFilePath) {
			this._logFilePath = logFilePath;
		} else {
			this.findLogFilePath();
		}

		this.setEndOfFile();
		this.watch();
	}

	private watch(): Promise<void> {
		return new Promise(async () => {
			while (true) {
				let newLines: string[] = [];

				do {
					newLines = await this.readNewLines();
				} while (newLines.length === 0);

				for (let line of newLines) {
					this.newLogEntry.next(line);
				}
			}
		});
	}

	private readNewLines(): Promise<string[]> {
		return new Promise((resolve) => {
			let lines: string[] = [];
			let currentPosition = this._endOfFile;
			this.setEndOfFile();

			if (currentPosition >= this._endOfFile) {
				return resolve(lines);
			}

			stat(this._logFilePath, (err, fileInfos) => {
				if (err) {
					console.error(err);
					return resolve(lines);
				}

				open(this._logFilePath, "r", (err, file) => {
					if (err) {
						console.error(err);
						return resolve(lines);
					}

					let buffer: Buffer;

					try {
						buffer = Buffer.alloc(fileInfos.size - this._endOfFile);
					} catch (e) {
						console.error(e);
						return resolve(lines);
					}

					read(
						file,
						buffer,
						0,
						buffer.length,
						currentPosition,
						(err, _, buffer) => {
							if (err) {
								console.error(err);
								return resolve(lines);
							}

							let linesStr: string = "";

							try {
								linesStr = buffer.toString();
							} catch (e) {
								console.error(e);
								return resolve(lines);
							}

							return resolve(linesStr.split(EOL));
						});
				});
			});
		});
	}

	private removeSpecialChars(str: string): string {
		return str ? str.replace(/[\\r\\n]/gi, "") : str;
	}

	private setEndOfFile(): Promise<void> {
		return Promise.resolve(stat(this._logFilePath, (err, infos) => {
			if (err) {
				console.info("Cannot get Client.txt stats.");
				return;
			}

			this._endOfFile = infos.size;
		}));
	}

	private async findLogFilePath(): Promise<void> {
		for (var name of this._poeProcesses) {
			var processes = await findProcess('name', name);

			if (processes && processes.length > 0) {
				const process = processes[0];
				const bin: string = (process as any).bin;

				if (platform() === 'win32') {
					this._logFilePath = bin.substring(0, bin.lastIndexOf("\\"));
					this._logFilePath += "\\logs\\Client.txt";
				} else {
					this._logFilePath = bin.substring(0, bin.lastIndexOf("/"));
					this._logFilePath += "/logs/Client.txt";
				}
			}
		}
	}
}